### PR TITLE
Demaion/raster2dggs

### PR DESCRIPTION
**Pull request by ndemaio**
This branch implements part of the backend for rhppandas by exposing some of the dggs functionality in a set of wrapper functions to the outside world. The structure and API of the new additions follow the example set by h3 to make them easy to use by someone familiar with h3.

Prominent features:
- New script rhp_wrappers.py: includes functions to convert from index to coordinates and vice versa, index operations like parent-child relationships, and a half-finished implementation of cell rings around a centre cell
- New script test_rhp_wrappers.py: provides test cases for the functions in rhp_wrappers.py

Not included:
- Polyfill and linefill
- Complex corner cases in cell rings - and, by extension, k-rings - remain unimplemented, which means that the results will be wrong in those situations (functions cell_ring and k_ring print a warning to that effect unless the verbose argument is set to False)

Other notes:
- Uses WSG84_003 as the default dggs where one is needed
- Assumes degrees not radians everywhere
- TODO comments in the function descriptions should point out missing functionality related to things that are available in the package but not exposed by the wrapper API

### CODE REVIEW

File: rhealpixdggs/rhp_wrappers.py [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+
+    Also returns None if k < 0.
+
+    Returns [ rhpindex ] if k == 0 (by convention).
+
+    Returns the four neighbouring faces at resolution 0 if k > 0 and cell resolution is 0
+    (by convention).
+    """
+    if verbose:
+        print(str.format(CELL_RING_WARNING, "cell"))
```
Comment (alpha-beta-soup) [inline rhealpixdggs/rhp_wrappers.py:276]: Should this rather use the built-in warnings module? https://docs.python.org/3/library/warnings.html#warnings.warn

File: rhealpixdggs/rhp_wrappers.py [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+def k_ring(rhpindex: str, k: int = 1, verbose=True) -> list[str]:
+    """
+    Returns the k-ring of cell indices around rhpindex at distance k (or None if rhpindex is invalid).
+
+    Also returns None if k < 0.
+
+    TODO: give the option to select another predefined DGGS, or pass in a custom one
+    """
+    if verbose:
+        print(str.format(CELL_RING_WARNING, "k"))
```
Comment (alpha-beta-soup) [inline rhealpixdggs/rhp_wrappers.py:372]: Ditto https://docs.python.org/3/library/warnings.html#warnings.warn

File: rhealpixdggs/rhp_wrappers.py (hunk 1 above)
Comment (ndemaio) [reply rhealpixdggs/rhp_wrappers.py]: You're absolutely right - I've changed that.

File: rhealpixdggs/rhp_wrappers.py (hunk 2 above)
Comment (ndemaio) [reply rhealpixdggs/rhp_wrappers.py]: Changed this here too.

**Comment by alpha-beta-soup** [comment]
I'll add some more review comments when I have more time. Great work though :)

File: rhealpixdggs/rhp_wrappers.py [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+
+def linetrace(geometry, resolution: int) -> list[str]:
+    raise NotImplementedError()
+
+
+# ======== Helper functions ======== #
+
+
+def _suid_to_str(suid: tuple) -> str:
+    return "".join([str(d) for d in suid])
```
Comment (alpha-beta-soup) [inline rhealpixdggs/rhp_wrappers.py:408]: The implementation differs from the `Cell.__str__` method, is that expected? It also seems to be used as `_suid_to_str(cell)`; would rather `str(cell)` be better?

File: tests/test_rhp_wrappers.py [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+
+# Import my modules
+import rhealpixdggs.rhp_wrappers as rhpw
+
+# Import helper modules
+import numpy as np
+
+
+class RhpWrappersTestCase(unittest.TestCase):
+    def test_geo_to_rhp(self):
```
Comment (alpha-beta-soup) [inline tests/test_rhp_wrappers.py:25]: Is it worth adding test cases for malformed/surprising input, perhaps to assert validation errors? Something like a |longitude| > 180?

**Comment by alpha-beta-soup** [comment]
I've added a few more tests. Of these, the second one fails, but not the first one:

```python
cell_id = rhpw.geo_to_rhp(11500249, 56898969, 4)
self.assertIsNone(cell_id) # passes

cell_id = rhpw.geo_to_rhp(270, 650, 4)
self.assertIsNone(cell_id) # fails, AssertionError: 'Q3333' is not None
```

File: rhealpixdggs/rhp_wrappers.py (hunk 3 above)
Comment (ndemaio) [reply rhealpixdggs/rhp_wrappers.py:408]: It's intentional that they're different. Cell.__str__ returns results that are not consistent across resolutions, which adds unnecessary complexity down the line. I feel that it's best to keep Cell.__str__ for display/print purposes only.

Passing the entire Cell object to _suid_to_str is worse from an encapsulation perspective: we'd be passing in a heavy object when we only need one tuple from its properties. The best practice that I learned is to separate out the tuple so the function doesn't need to know about the Cell class as a whole. (That also means that e.g. the dggs object could pass in an appropriate tuple id and have it converted to string - if that's something we need in the future.)

File: tests/test_rhp_wrappers.py (hunk 4 above)
Comment (ndemaio) [reply tests/test_rhp_wrappers.py:25]: The underlying function that implements the conversion is `RHEALPixDGGS.cell_from_point`. It does have its own tests in `test_dggs.py` and any additions should probably go in there rather than in the wrapper tests.

Coordinates that don't lie in the dggs will result in None being returned, which is passed up through the wrapper (and documented in the function documentation for the wrapper).

I have vague memories of trying to find a test case with invalid coordinates and failing - it's possible that longitude values wrap around the sphere?

**Comment by alpha-beta-soup** [comment]
OK, I understand this a bit better now. I've made effectively the same test twice, once in the wrapper and once for `RHEALPixDGGS.cell_from_point`. Both assert that `None` is returned given a point that is not on the planar DGGS, which is what I would expect.

File: rhealpixdggs/rhp_wrappers.py (hunk 3 above)
Comment (alpha-beta-soup) [reply rhealpixdggs/rhp_wrappers.py:408]: > Passing the entire Cell object to _suid_to_str

I'm not recommending this; but it does seem odd looking in that there's a cell `__str__` method which does almost the same thing, but we don't use it in favour of this private function. The difference is only apparent when the condition `(self.rdggs.N_side) ** 2 < 10` is met, and given we're only using `N_side = 3`, this condition won't be met. Are we therefore better off using `Cell.__str__` for the sake of consistency?

Comment (ndemaio) [reply rhealpixdggs/rhp_wrappers.py:408]: We can use that if you're happy to make the changes to the wrapper. I'm a bit out of hours to dive into it atm.

Comment (alpha-beta-soup) [reply rhealpixdggs/rhp_wrappers.py]: I've removed the function in favour of using Cell.__str__; no other changes needed and all tests pass.
